### ISSUE TITLE
How to stop auto indent?

### ISSUE THREAD

**Issue by fuzihaofzh** [syl20bnr/spacemacs#4219]
I want to copy a pice of python code into emacs. But emacs makes it automatically to indent to make my code a mess. How can I solve the problem?

**Comment by tuhdo** [comment]
You can use prefix argument by pressing `C-u` before pasting with `C-y`. If you are using evil-mode, pressing `SPC u` before pressing `p`.

**Comment by syl20bnr** [comment]
The auto-indent should not happen in python buffers so I suppose you are pasting in a buffer which is not a python buffer. Is it the case ? If yes then @tuhdo answer is the way to go, if no then we have a bug.

**Comment by maplewizard** [comment]
I use emacs in iterm to login in a remote client. so I just use command+v to paste in iterm.

**Comment by puffnfresh** [comment]
Auto-indenting is very annoying with Nix code. I've tried using `SPC u p` but it has never worked for me.

**Comment by magnetophon** [comment]
```SPC u p``` doesn't work in Nix code.

#### System Info
- OS: gnu/linux
- Emacs: 24.5.2
- Spacemacs: 0.105.4
- Spacemacs branch: master (rev. df8eb1b)
- Distribution: spacemacs
- Layers:
```elisp
(auto-completion erc emacs-lisp extra-langs eyebrowse fasd git github markdown nixos org ranger search-engine spell-checking syntax-checking version-control)
```

**Comment by dlukes** [comment]
> I use emacs in iterm to login in a remote client. so I just use command+v to paste in iterm.

+ 1, I experience the same behavior -- pasting in a Python buffer results in the contents of the clipboard being auto-indented beyond recognition.

**Comment by shakerlxxv** [comment]
+1 I'm pasting a bit of SQL into a SQL file and seeing similar "flattened" results after pasting. I'm in evil mode, trying `SPC-u p` doesn't change the result.

**Comment by aldanor** [comment]
How do I do a "raw paste" in prog-mode? 

`SPC-u p` just results in `Wrong type argument: number-of-marker-p, (4)` error.

**Comment by laMudri** [comment]
I seem to have the same problem in Agda and Nix code. Any pasted lines are indented to the indentation level of the previous non-empty line. Python does no indentation (as wanted).

**Comment by aldanor** [comment]
Do any spacemacs gurus know the answer? :wink:

**Comment by yashi** [comment]
Just curious.  I don't use spacemacs nor i-something, but are you all using spacemacs under some terminal emulator?  And copy from something and past it on the terminal?  If so, emacs can't distinguish _pasted enter_ and _key input enter_ and electric-indent kicks in.

What you can do is to disable electric indent with `M-x electric-indent-mode`. Past whatever you want, and do `M-x electric-indent-mode` again to re-enable it.

Hope this helps.

**Comment by laMudri** [comment]
@yashi, no, I'm using graphical Emacs (GNU/Linux). I'm also using Evil's paste, rather than sending keystrokes, so I would be surprised if that were the problem. The problem also occurs with `electric-indent-mode` disabled.

**Comment by aldanor** [comment]
I'm also using graphical Emacs (Mac). Can confirm, the problem occurs with electric indent disabled too.

**Comment by eskimor** [comment]
Same problem here, also GUI, disabling electric indent does not work. Switching to fundamental mode from nix-mode, does help though! :-) yeah!

**Comment by mosic** [comment]
I encountered this problem today when working in `elixir-mode`.  I found the item in the docs which describes how to change the pasting behaviour: [DOCUMENTATION.org#auto-indent-pasted-text](https://github.com/syl20bnr/spacemacs/blob/master/doc/DOCUMENTATION.org#auto-indent-pasted-text)

This is what I added to my `.spacemacs` file to make it work:

``` elisp
(add-to-list 'spacemacs-indent-sensitive-modes 'elixir-mode)
```

**Comment by clhodapp** [comment]
https://github.com/syl20bnr/spacemacs/blob/5f26b82e1abdde81cdf7cd17ba06f64db2343667/layers/%2Bdistribution/spacemacs-base/packages.el#L337

https://github.com/syl20bnr/spacemacs/blob/5f26b82e1abdde81cdf7cd17ba06f64db2343667/layers/%2Bdistribution/spacemacs-base/funcs.el#L834

`Spc-u 4 p` actually does paste four times without autoindention. Fun.

**Comment by clhodapp** [comment]
If you change the the second one to,` (if (and (not (equal nil (ad-get-arg 0))) ...)` then it works.

**Comment by clhodapp** [comment]
Proposal replace both with a single working version:
```
(spacemacs|advise-commands
 "indent" (yank yank-pop evil-paste-before evil-paste-after) around
 "If current mode is not one of spacemacs-indent-sensitive-modes
 indent yanked text (with universal arg don't indent)."
 (let ((prefix (ad-get-arg 0)))
	 (ad-set-arg 0 (unless (equal '(4) prefix) prefix))
	 (evil-start-undo-step)
	 ad-do-it
	 (if (and (not (equal '(4) prefix))
						(not (member major-mode spacemacs-indent-sensitive-modes))
						(or (derived-mode-p 'prog-mode)
								(member major-mode spacemacs-indent-sensitive-modes)))
			 (let ((transient-mark-mode nil)
						 (save-undo buffer-undo-list))
				 (spacemacs/yank-advised-indent-function (region-beginning)
																								 (region-end))))
	 (evil-end-undo-step)
	 (ad-set-arg 0 prefix)))
```

**Comment by Blaisorblade** [comment]
With `SPC u p` I also get the following warnings:

```
ad-Advice-evil-paste-after: Wrong type argument: number-or-marker-p, (4) [2 times]
evil-yank-line-handler: Wrong type argument: wholenump, (4)
```

**Comment by Blaisorblade** [comment]
@clhodapp Since you seem to have found a solution, is there any chance you could send a pull request?

**Comment by aviggiano** [comment]
I'm still having this issue
@mosic Could you please post your dotfile with the appropriate code for the SQL layer? I tried to generalize your elixir piece of code but it didn't work

**Comment by henriquenj** [comment]
@aviggiano You can disable auto-indent on <kbd>p</kbd> altogether using the `spacemacs-indent-sensitive-modes` variable.  You can select this behaviour on a per-mode basis. If you think that is good for your workflow, here is the code for `sql-mode`:
```elisp
(add-to-list 'spacemacs-indent-sensitive-modes 'sql-mode)
```
Just put into your `dotspacemacs/user-config`

**Comment by duianto** [comment]
This PR has been merged:
Check for universal argument before paste #11340

It superseeded this PR:
Fix pasting with universal argument #11858
which is marked as fixing this issue.
